rclpy/rclpy/parameter_event_handler.py

```diff
@@ -0,0 +1,208 @@
+
+                for event_callback in reversed(self.event_callbacks):
+                    with event_callback.mutex:
+                        event_callback.callback(event)
+
+        def add_parameter_callback(
+            self,
+            parameter_name: str,
+            node_name: str,
+            callback: Callable[[Parameter], None],
+        ) -> ParameterCallbackHandle:
+            handle = ParameterCallbackHandle(
+                parameter_name=parameter_name,
+                node_name=node_name,
+                callback=callback,
+            )
+
+            with self.mutex:
+                self.parameter_callbacks[(parameter_name, node_name)].append(handle)
+
+            return handle
+
+        def remove_parameter_callback(
+            self,
+            handle: ParameterCallbackHandle,
+        ) -> None:
+            with self.mutex:
+                handle_key = (handle.parameter_name, handle.node_name)
+                if handle_key in self.parameter_callbacks:
+                    if handle in self.parameter_callbacks[handle_key]:
+                        self.parameter_callbacks[handle_key].remove(handle)
+                    else:
+                        raise RuntimeError("Callback doesn't exist")
+
+                    if len(self.parameter_callbacks[handle_key]) == 0:
+                        self.parameter_callbacks.pop(handle_key)
+                else:
+                    raise RuntimeError("Callback doesn't exist")
+
+        def add_parameter_event_callback(
+            self,
+            callback: Callable[[ParameterEvent], None],
+        ) -> ParameterEventCallbackHandle:
+            handle = ParameterEventCallbackHandle(callback=callback)
+
+            with self.mutex:
+                self.event_callbacks.append(handle)
+
+            return handle
+
+        def remove_parameter_event_callback(
+            self,
+            handle: ParameterEventCallbackHandle,
+        ) -> None:
+            with self.mutex:
+                if handle in self.event_callbacks:
+                    self.event_callbacks.remove(handle)
+                else:
+                    raise RuntimeError("Callback doesn't exist")
+
+    def __init__(
+        self,
+        node: Node,
+        qos_profile: QoSProfile = qos_profile_parameter_events,
+        callback_group: Optional[CallbackGroup] = None,
+        event_callbacks: Optional[SubscriptionEventCallbacks] = None,
+        qos_overriding_options: Optional[QoSOverridingOptions] = None,
+        raw: bool = False,
+    ):
+        """
+        Create ParameterEventHandler.
+
+        A class used to "handle" (monitor and respond to) changes to parameters.
+        :param node: Used to subscribe to parameter_events topic
+        """
+        self.node = node
+        self.qos_profile = qos_profile
+
+        self._callbacks = ParameterEventHandler.Callbacks()
+
+        self.parameter_event_subscription = node.create_subscription(
```

Comment:
I think we need to add in a `destroy` method that explicitly undoes the subscription.

We need that kind of thing in Python because we can't be sure when the garbage collector will come along and destroy this object, and we don't always want to wait around for that.  So we have explicit `destroy` methods to allow the user to control that.  In this case it will just do a `node.destroy_subscription`.